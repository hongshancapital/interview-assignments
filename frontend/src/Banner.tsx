import React, {Children} from 'react';

interface Props {
    slider: number;
    total: number;
    children: JSX.Element[] | JSX.Element;
}

const Banner = ({
    slider,
    total,
    children,
}: Props) => (
    <div 
        className="wrap"
        style={{
            width: `${total * 100}%`,
            transform: `translateX(-${100 * slider / total}%)`,
        }}
    >
        {Children.map(children, child => (
            <div 
                className="image-item"
                style={{
                    width: `${100 / total}%`,
                }}
            >{child}</div>))}
    </div>
);

export default Banner;