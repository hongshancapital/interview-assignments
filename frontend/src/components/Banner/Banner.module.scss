.banner {
    height: 100%;
    background-repeat: no-repeat;
    background-size: auto 500px;
    background-position: 50% 79%;
}

.slogan {
    padding-top: 18%;
    height: 120px;

    h1 {
        margin-top: 0;
        margin-bottom: 12px;
        text-align: center;
        font-size: 72px;
        white-space: normal;
        word-wrap: normal;
    }

    p {
        margin: 0;
        font-size: 36px;
    }
}