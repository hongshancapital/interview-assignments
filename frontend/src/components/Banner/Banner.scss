$banner-prefix: 'banner';
.#{$banner-prefix}-wrapper{
    display: flex;
    align-items: flex-start;
    flex-direction: column;
    align-items: center;
    position: relative;
    height: 100%;
    .#{$banner-prefix}-info{
        display: flex;
        flex-direction: column;
        margin-top: 10%;
    }
    .#{$banner-prefix}-title{
        font-size: 50px;
        white-space: pre-wrap;
        z-index: 1;
    }
    .#{$banner-prefix}-desc{
        font-size: 28px;
        white-space: pre-wrap;
        z-index: 1;
    }
    .#{$banner-prefix}-img {
        width: 200px;
        height: 350px;
        object-fit: cover;
        object-position: bottom;
        position: absolute;
        z-index: 0;
        bottom: 8%;
    }
}