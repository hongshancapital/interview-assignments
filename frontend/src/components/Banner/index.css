.banner {
  height: 100%;
  position: relative;
  background-repeat: no-repeat;
  background-size: auto 50%;
  background-position: center 70%;
  white-space: pre-wrap;
  padding-top: 25%;
  box-sizing: border-box;
}
.banner-content {
  position: absolute;
  top: 25%;
  width: 100%;
  text-align: center;
}
