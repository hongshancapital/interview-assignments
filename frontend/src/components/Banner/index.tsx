import { memo, FC } from 'react';
import './index.css';

interface BannerProps {
  /** carousel classname */
  className?: string;
  /** banner related data */
  data: {
    /** the title of the banner element */
    title: string;
    /** detailed description of the banner element */
    text?: string;
    /** the background image of the banner element */
    img?: string;
    /** the font color of the banner element */
    color?: string;
    /** the background color of the banner element */
    backgroundColor?: string;
  };
}

const Banner: FC<BannerProps> = function ({ className = '', data }) {
  const { title, text, color, backgroundColor, img } = data;
  return (
    <div
      className={`banner ${className}`}
      style={{
        color,
        backgroundColor,
        backgroundImage: img ? `url(${img})` : undefined,
      }}
    >
      <div className="banner-content">
        <h3 className="title">{title}</h3>
        {text && <p className="text">{text}</p>}
      </div>
    </div>
  );
};

export default memo(Banner);
