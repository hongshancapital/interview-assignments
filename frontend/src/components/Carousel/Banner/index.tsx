import { CSSProperties, FC, useMemo } from "react";
import "./index.css";

/** 主题，用于控制文本颜色 */
export enum Them_Enum {
  dark = "dark", // 暗夜
  light = "light", // 明亮
}

/**
 * 轮播内容
 */
export interface BannerProp_Inter {
  /** 主图路径 */
  imgUrl: string;
  /** 背景色 */ 
  backgroundColor: string;
  /** 标题列表 */
  title: string[];
  /** 主色调(影响文本颜色,避免深色文本在深色背景上不易读) */
  them?: Them_Enum;
  /** 文本内容(小标题) */
  text?: string[];
}

type BannerProp = BannerProp_Inter & {
  index: number;
};

const Banner: FC<BannerProp> = ({
  imgUrl,
  title,
  text,
  index,
  them = Them_Enum.light,
  backgroundColor,
}) => {
  const bannerStyle: CSSProperties = useMemo(
    () => ({
      backgroundColor,
      backgroundImage: `url(${imgUrl})`,
      position: "absolute",
      top: 0,
      left: `${index}00%`,
    }),
    [index, imgUrl, backgroundColor]
  );

  const titleDOM = useMemo(
    () =>
      title.map((titleContent, index) => (
        <h1 className="title" key={index}>
          {titleContent}
        </h1>
      )),
    [title]
  );

  const textDOM = useMemo(
    () =>
      text?.map((textContent, index) => (
        <p className="text" key={index}>
          {textContent}
        </p>
      )),
    [text]
  );

  return (
    <div
      className={`banner_control ${them === Them_Enum.dark && Them_Enum.dark}`}
      style={bannerStyle}
    >
      {titleDOM}
      {textDOM}
    </div>
  );
};

export default Banner;
