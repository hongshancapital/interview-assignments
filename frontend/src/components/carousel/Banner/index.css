.banner_control{
  width: 100%;
  height: 100%;
  padding-top:210px;
  box-sizing: border-box;
  background-size: auto 133px;
  background-repeat: no-repeat;
  background-position: center 71.6%;
  position: absolute;
}

.banner_control .title{
  margin-left: auto;
  margin-right: auto;
}

.banner_control .text{
  margin-left: auto;
  margin-right: auto;
}

.banner_control.dark .title,
.banner_control.dark .text{
  color:#fff;
}