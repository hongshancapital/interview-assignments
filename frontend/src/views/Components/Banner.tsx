import React,{FunctionComponent} from 'react'
import {TitleFont,TextFont,Img,Wrapper} from '../../styles'

type Props = {
  title:string;
  subTitle?:string;
  img:string;
  theme?:string|'dark'|'light';
}

const App:FunctionComponent<Props> = (props) => {
  return (
    <Wrapper {...props}>
      <TitleFont>{props.title}</TitleFont>
      <TextFont>{props.subTitle}</TextFont>
      <Img src = {props.img}/>
    </Wrapper>
  )
}

export default App
