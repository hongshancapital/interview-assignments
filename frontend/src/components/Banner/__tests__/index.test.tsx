import { render } from '@testing-library/react';
import Banner from '../index';
describe('test Banner component', () => {
  const data = {
    title: 'xPhone',
    text: 'Lots to love.Less to spend.\nStarting at $399.',
    color: '#fff',
    img: 'test img',
    backgroundColor: '#101010',
  };

  it('renders correctly', () => {
    const { container } = render(<Banner data={data} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('mount correctly', () => {
    expect(() => render(<Banner data={data} />)).not.toThrow();
  });

  it('test rendering properties', () => {
    const { container } = render(<Banner className="test" data={data} />);

    expect(container.querySelector('.banner')).toHaveClass('test');
    expect(container.querySelector('.banner')).toHaveStyle({
      color: '#fff',
      backgroundColor: '#101010',
    });

    expect(container.querySelector('h3')?.textContent).toBe('xPhone');

    expect(container.querySelector('p')?.textContent).toBe(
      'Lots to love.Less to spend.\nStarting at $399.'
    );
  });
});
