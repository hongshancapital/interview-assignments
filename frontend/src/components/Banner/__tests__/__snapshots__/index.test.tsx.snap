// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner renders correctly 1`] = `
<div
  class="banner "
  style="color: rgb(255, 255, 255); background-color: rgb(16, 16, 16);"
>
  <div
    class="banner-content"
  >
    <h3
      class="title"
    >
      xPhone
    </h3>
    <p
      class="text"
    >
      Lots to love.Less to spend.
Starting at $399.
    </p>
  </div>
</div>
`;

exports[`test Banner component renders correctly 1`] = `
<div
  class="banner "
  style="color: rgb(255, 255, 255); background-color: rgb(16, 16, 16);"
>
  <div
    class="banner-content"
  >
    <h3
      class="title"
    >
      xPhone
    </h3>
    <p
      class="text"
    >
      Lots to love.Less to spend.
Starting at $399.
    </p>
  </div>
</div>
`;
